### ISSUE TITLE
Runtime error: Generated Kernel is NULL

### ISSUE THREAD

**Issue by vtschouten97** [beehive-lab/TornadoVM#20]
Hello TornadoVM team!

I'm in the process of implementing a branch-and-bound algorithm in TornadoVM.
It seems like all my TaskSchedules report that the generated kernel is null, and I haven't found a way to resolve it yet.

The first TaskSchedule is called as follows:
new TaskSchedule("Internal-UB-SV13-Calculation") //
.streamIn(ratios) //	.task("taskSV13",ParallelBBModel::loopInternalSV13,internalUBs,ratios,currWeight,currProfit,currSize,capacity) //
.streamOut(internalUBs) //
.executeWithProfiler(Policy.PERFORMANCE);

(Part of) the error message is as follows:
uk.ac.manchester.tornado.api.exceptions.TornadoRuntimeException: [ERROR] Generated Kernel is NULL. 
Please report this issue to https://github.com/beehive-lab/TornadoVM
at uk.ac.manchester.tornado.drivers.opencl.graal.OCLInstalledCode.checkKernelNotNull(OCLInstalledCode.java:353)
at uk.ac.manchester.tornado.drivers.opencl.graal.OCLInstalledCode.submitWithoutEvents(OCLInstalledCode.java:359)
at uk.ac.manchester.tornado.drivers.opencl.graal.OCLInstalledCode.launchWithoutDependencies(OCLInstalledCode.java:389)
at uk.ac.manchester.tornado.runtime.TornadoVM.execute(TornadoVM.java:507)
at uk.ac.manchester.tornado.runtime.TornadoVM.execute(TornadoVM.java:197)
at uk.ac.manchester.tornado.runtime.tasks.TornadoTaskSchedule.scheduleInner(TornadoTaskSchedule.java:436)
at uk.ac.manchester.tornado.runtime.tasks.TornadoTaskSchedule.schedule(TornadoTaskSchedule.java:666)
at uk.ac.manchester.tornado.api.TaskSchedule.execute(TaskSchedule.java:230)
at uk.ac.manchester.tornado.runtime.tasks.TornadoTaskSchedule.lambda$runParallelTaskSchedules$1(TornadoTaskSchedule.java:919)
at java.lang.Thread.run(Thread.java:748)
Bailout from LAUNCH Bytecode: 
Reason: uk.ac.manchester.tornado.api.exceptions.TornadoRuntimeException: [ERROR] Generated Kernel is NULL. 
Please report this issue to https://github.com/beehive-lab/TornadoVM
uk.ac.manchester.tornado.runtime.TornadoVM.execute(TornadoVM.java:517)
uk.ac.manchester.tornado.runtime.TornadoVM.execute(TornadoVM.java:197)
uk.ac.manchester.tornado.runtime.tasks.TornadoTaskSchedule.scheduleInner(TornadoTaskSchedule.java:436)
uk.ac.manchester.tornado.runtime.tasks.TornadoTaskSchedule.schedule(TornadoTaskSchedule.java:666)
uk.ac.manchester.tornado.api.TaskSchedule.execute(TaskSchedule.java:230)
uk.ac.manchester.tornado.runtime.tasks.TornadoTaskSchedule.lambda$runParallelTaskSchedules$1(TornadoTaskSchedule.java:919)
	java.lang.Thread.run(Thread.java:748)
task info: XXX0.taskSV13
	platform          : NVIDIA CUDA
	device            : GeForce RTX 2060 CL_DEVICE_TYPE_GPU (available)
	dims              : 1
	global work offset: [0]
	global work size  : [250]
	local  work size  : [250, 1, 1]
uk.ac.manchester.tornado.drivers.opencl> notify error:
uk.ac.manchester.tornado.drivers.opencl> CL_INVALID_KERNEL_ARGS error executing CL_COMMAND_NDRANGE_KERNEL on GeForce RTX 2060 (Device 0).

I do hope you could shed some light on the culprit.

Kind regards,
Victor Schouten

**Comment by jjfumero** [comment]
Can you provide the `loopInternalSV13` method? 

It looks to me there is a problem to get the final OpenCL code/ 

Also, can you report the generated kernel? You can use `--printKernel` 

There is also an option `-Dtornado.fullDebug=True` that enables all Debug info, including the JIT compiler and runtime traces.

**Comment by stratika** [comment]
Hi @vtschouten97, have you tried the unit-tests by running `make tests`? If they work, then the problem must be related to the method that you pass in your TaskSchedule, or the parameters in your TaskSchedule. As @jjfumero mentioned, it would be nice to have some information about the method that you compile and the command that you run your program.

**Comment by vtschouten97** [comment]
A couple of weeks ago I did run a unit-test and it seemed to work fine. But now that you mention it, I've tried to run the tests again and it' s not passing all the tests. It's failing for the majority of them...

**Comment by jjfumero** [comment]
Can you recompile TornadoVM with the latest version in the `master` branch? Also, can you report the `tornado --devices` ? Minimum required is a device with OpenCL 1.2

**Comment by vtschouten97** [comment]
After recompiling with newly cloned directory of TornadoVM, tornado -- devices shows this:

Number of Tornado drivers: 1
Total number of devices  : 2
Tornado device=0:0
	NVIDIA CUDA -- GeForce RTX 2060
		Global Memory Size: 5.8 GB
		Local Memory Size: 48.0 KB
		Workgroup Dimensions: 3
		Max WorkGroup Configuration: [1024, 1024, 64]
		Device OpenCL C version: OpenCL C 1.2

Tornado device=0:1
	Intel Gen OCL Driver -- Intel(R) HD Graphics IvyBridge GT2
		Global Memory Size: 2.0 GB
		Local Memory Size: 64.0 KB
		Workgroup Dimensions: 3
		Max WorkGroup Configuration: [512, 512, 512]
		Device OpenCL C version: OpenCL C 1.2 beignet 1.3

Is it normal there to be warning messages when calling make graal-jdk-8?
Other information:
python3 --version: python 3.8.2
gcc --version: gcc (Ubuntu 9.3.0-10ubuntu2) 9.3.0
mvn --version: Apache Maven 3.6.3
cmake --version: cmake version 3.16.3

Unit-tests still seem to fail

**Comment by jjfumero** [comment]
When running the unittests, the file `tornado_unittests.log` is generated with all the report. Can you share this file? 

If you run `make graal-jdk-8`, your JDK should be `graalvm-ce-java8-19.3.0` 
Also Python 2.7 is required.

**Comment by vtschouten97** [comment]
I've installed graalvm-ce-java8-19.3.0 as explained in the installation guide.
Pyhton should be okay now:
python2.7 --version: Python 2.7.18
I've added the unit-test log file.

[tornado_unittests.log](https://github.com/beehive-lab/TornadoVM/files/4772132/tornado_unittests.log)

**Comment by jjfumero** [comment]
@vtschouten97 this is very strange. All unittests should pass.

 Which OS are you running? 
The master branch has been tested on:
* Ubuntu 18, 20
* CentOS 7.4 - 7.8 
* Mac OSx

Since the unittests are not passing, I think your issue is related to the TornadoVM configuration. 

BTW, CMake has to be >= 3.6 (See pre-requisites here) https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/10_INSTALL_WITH_GRAALVM.md#installing-tornadovm-with-graalvm-1930)

Here my list of variables needed to run TornadoVM:

```bash
## ###############################################
## CUDA - OpenCL
## ###############################################
export CPLUS_INCLUDE_PATH=/usr/local/cuda/include
export C_INCLUDE_PATH=/usr/local/cuda/include
export LD_LIBRARY_PATH=/usr/local/cuda/lib64

## ###############################################
## CMAKE 3.10
## ###############################################
export PATH=/home/juan/bin/cmake/cmake-3.10.2/bin:$PATH

## ###############################################
## TornadoVM
## ###############################################
export PATH=$PWD/bin/bin:$PATH
export TORNADO_SDK=$PWD/bin/sdk
export CMAKE_ROOT=/home/juan/bin/cmake/cmake-3.10.2
export JAVA_HOME=/home/juani/graal-19.3/graal-jvmci-8/openjdk1.8.0_242/linux-amd64/product
```


Alternatively, you can use the docker images: https://github.com/beehive-lab/docker-tornado 

If you want to run on Nvidia GPUs with Graal CE JDK8 you can run. assuming you have the NVIDIA Docker Daemon already running (https://github.com/NVIDIA/nvidia-docker). 

```bash
$ docker pull beehivelab/tornado-gpu-graalvm-jdk8:latest 
```

**Comment by vtschouten97** [comment]
I'm running Ubuntu 20.04 LTS. As mentioned before, my cmake is version 3.16.3, so that shouldn't be the problem. I assume the above code belongs in the sources.env file. The commands regarding cuda are not mentioned in the installation guide. 

The curious thing is I was able to perform a benchmark some weeks ago and it showed me the speedup TornadoVM had realised using the GPU. I think I'll try to do a clean installation of Ubuntu and get TornadoVM up and running again, that seems like the best course of action for me know I think.

**Comment by stratika** [comment]
It seems to be a mis-configuration with the settings, the unit-tests should not fail. I would try to install tornado and graal from scratch, instead of directly installing Ubuntu again. 

To clarify your questions, the variables are in the sources.env file and the cuda paths are not required.

**Comment by vtschouten97** [comment]
I will tinker around a bit in the coming days and will let you know what the progress is. I thank you all for the swift responses!

**Comment by vtschouten97** [comment]
I've had a bit of struggle getting Ubuntu to run as I want it to, but it's working now. Unit tests all pass and my code does not give errors now. It does seem though that at one point the program just stops running, yet it does not give an error. Does this maybe have something to do with heap memory or the like?

This is what is displayed when I run my sequential code:
Loading dataset...
Dataset loaded in 0.36 seconds

Commencing solving procedure...
Solving procedure finished.
Amount of instances solved: 1
Finished solving at 17:50:59.913
Commencing writing to file now...
Results written to file.

With GPU implementation in TornadoVm, it shows this:
Loading dataset...
Dataset loaded in 0.36 seconds
Commencing solving procedure...
victor@victor-desktop:~/Desktop/ParallelBranch$

At least TornadoVM is now working as it should, that gives me an opportunity to play with my implementation of the branch-and-bound algorithm. Thanks for the help, I'll probably come knocking again when I've hit a wall haha

**Comment by jjfumero** [comment]
@vtschouten97 , a side note. The number of task-schedules to be instantiated per application is limited to the amount of memory available on the target device. We are working on it to print this in the debug information. If possible, try to reuse task-schedules. You can copy new data by using the calls `streamIn` and `streamOut`.

**Comment by vtschouten97** [comment]
Hello all, I am back again with a small question. 

I've added the code to make my question more clear. The TaskSchedule is working, but when I try to reuse it it doesn't seem to update some input/output. I've played around with streamIn and streamOut but I keep getting the same results. The code is quite simplistic, but I'll elaborate a bit. There is a loop in which the TaskSchedule is executed, and I need to use the loopcounter _k_ in the TaskSchedule. Then in the TaskSchedule, the array inputDP is used only for reading, while the array outputDP is modified. In each consecutive iteration it needs to have the updated version of both arrays, yet after some debugging that does not seem the case.

Any help on this matter would be greatly appreciated!

Kind regards, 
Victor

[boundCompSnippet.zip](https://github.com/beehive-lab/TornadoVM/files/4814545/boundCompSnippet.zip)

**Comment by stratika** [comment]
Hi, the code seems to be compiling correctly. As a first remark, I would say that `streamIn` and `streamOut` do not need to be invoked in every iteration but it should not be a problem. 

I have a question, if you run it once, does it return the correct results?

It would be better if you could provide with your main method and the exact test you run, in order to be able to reproduce your case.

Best Regards,
Thanos

**Comment by vtschouten97** [comment]
It's a project containing multiple source files, as requested I've everything needed to run it. The java class of interest is ParallelBBModel.java, where the code snippet I posted can be found at line 114, and the function is called at line 87. The Main class should be called to run the project, after which you'll get a couple of prompts. Concerning the dataset, enter R1; then for the branch-and-bound model, enter 4 (the BCM17 variant). When it runs it will display the sequential (correct) value and the parallel value. There are some external libraries included in the folder _lib_, as it needs to write data to an Excell file. Thanks again for the quick response!  

[parallel_git.zip](https://github.com/beehive-lab/TornadoVM/files/4814732/parallel_git.zip)

**Comment by stratika** [comment]
Can you please provide some information how do you run the program? what are the input arguments? I see a menu that asks for information but there is only one file in the data directory with name `BPPC_1_0_1.txt_0.1.`

**Comment by vtschouten97** [comment]
It's a Java project for a branch-and-bound algorithm. The data file is unimportant for the TaskSchedule, it merely contains a problem instance for a Knapsack Problem with Conflict Graph. The menu is to choose a specific branch-and-bound algorithm, as there are multiple implementations. If you follow the instructions I posted above when running the main java class, you'll end up in the ParallelBranchBound class. At line 87 the function internalBCM17 is called, which contains the TaskSchedule (line 114). In a previous message I remarked two arrays, one to be read and one to be changed (inputDP and outputDP). Besides that there are arrays containing profits and weights for the items, an integer capacity, and integer k which depends on the loop. The method parallelInternalBCM17 at line 136 is to be executed on the parallel device. Method seqTestInternalBCM17 is a sequential implementation of parallelInternalBCM17.

The following lines of code is how I compile and run the code, given that the external libraries are imported.
javac.py parallel/Main.java parallel/Item.java parallel/Instance.java parallel/InstanceLoaderjava parallel/Clique.java parallel/CliqueHeuristic.java parallel/ParallelBBModel.java parallel/ExcelWriter.java 
tornado parallel/Main

I hope this is the information you need. If there are any more things, let me know.

Kind regards,
Victor

**Comment by stratika** [comment]
Hi Viktor, 

At first I had a problem with the relative path that in my laptop resulted in `NullPointerException` for the listed files in the directory.

Then, I noticed that the menu option and the second input from the terminal matter. Especially, the second input as the first choice: _**- 0 for SV13**_, is executed ok and it does not seem to invoke the method that is forwarded to TornadoVM. 

In fact, there is a condition for `preKP`, so then I opted the next choice _**- 1 for SV13 with preKP**_.

After doing this, I get the following error, in a MacBook Pro and particularly the Intel(R) Core(TM) i5-5257U CPU device: 
`
uk.ac.manchester.tornado.drivers.opencl> notify error:
uk.ac.manchester.tornado.drivers.opencl> [CL_INVALID_ARG_INDEX] : OpenCL Error : clSetKernelArg failed: arg_index = 4 is too big. This kernel only has 4 arguments.
`

Is this the problem that you get? 

Also when I tried `ts.execute()` instead of `ts.executeWithProfiler(Policy.PERFORMANCE)`, I got several messages indicating errors at the TornadoVM compiler and in particular in the Low IR phase. The message looks like this: 

`java.lang.NullPointerException
java.lang.NullPointerException
	at org.graalvm.compiler.core.common.LIRKind.combine(LIRKind.java:182)
	at uk.ac.manchester.tornado.drivers.opencl.graal.lir.OCLArithmeticTool.emitDiv(OCLArithmeticTool.java:135)`

Please confirm that this is the problem, because I thought that you had a problem with wrong results. Also, is the CUDA device still your device 0:0? Could you try to execute on a different device with the following flags: `-D<taskScheduleName>.<taskName>.device=0:1` and `--debug`.

**Comment by vtschouten97** [comment]
Eventhough it is becoming quite late now, I've extracted the following from my code. I would like to say, I have not encountered any NullPointerException in my endeavours.

This is the first message which came up when ts.execute() was called.
ERROR] clEnqueueReadBuffer, code = -5
uk.ac.manchester.tornado.drivers.opencl> notify error:
uk.ac.manchester.tornado.drivers.opencl> CL_OUT_OF_RESOURCES error executing CL_COMMAND_READ_BUFFER on GeForce RTX 2060 (Device 0).

which is a common I noticed when looking at the reports when calling --debug

However when executing ts.executeWithProfiler(Policy.PERFORMANCE) it conveys the following message:
Commencing solving procedure...
Sequential: 405.0 - Parallel: 99.0
Sequential: 405.0 - Parallel: 183.0
Sequential: 402.0 - Parallel: 244.0
Sequential: 402.0 - Parallel: 329.0
Sequential: 402.0 - Parallel: 375.0
Sequential: 375.0 - Parallel: 375.0
Sequential: 402.0 - Parallel: 402.0
Sequential: 402.0 - Parallel: 402.0
Sequential: 394.0 - Parallel: 329.0
Sequential: 389.0 - Parallel: 290.0
Sequential: 387.0 - Parallel: 268.0
Sequential: 379.0 - Parallel: 175.0
Sequential: 371.0 - Parallel: 84.0
402.0
Solving procedure finished.

_preKP_ is indeed the condition for the parallel implementation to take place. Choice 4 is just generally faster, hence why I recommended its use. With the performance policy, as I've shown now, there is a discontinuity between sequential and parallel values.

Kind regards,
Victor

**Comment by stratika** [comment]
Hi Viktor, I will have a closer look today and will email you to arrange a chat.

**Comment by jjfumero** [comment]
We are not able to reproduce the error, as you described. I suggest providing a **simple test-case** that is simpler to test and to spot the bug. 

One issue we found (not relevant to TornadoVM) is that you are launching kernels on the GPU with ~70 threads. Try to increase the number of threads to run. 

We also found another set of problems in TornadoVM from this use case:
* Kernel is serialized when the input application is not array-based. 
* Use of stream-in after the task definition (task-definition triggers a sketch-compilation with data type specialization). 

We will work on those issues.

**Comment by vtschouten97** [comment]
With some tips from Thanos I've adjusted the code a bit, unfortunately something is still going wrong.
I've made a simple test-case, which I've included in this message.
To clarify the method, it is a dynamic programming algorithm for the 0-1 knapsack problem.
The test-case prints for each item iteration the TaskSchedule output and sequential output.
There it shows that for the first iteration the computations are done correctly, but the following iterations the two outputs diverge.
I've included the log of the TornadoVM unit-tests as well, just in case.

The debug flag reports the following
task info: Inner-Loop-Internal-UB-BCM17.taskBCM17-0
	platform          : NVIDIA CUDA
	device            : GeForce RTX 2060 CL_DEVICE_TYPE_GPU (available)
	dims              : 0
	global work offset: [0]
	global work size  : [1]
	local  work size  : null

[testcase_git.zip](https://github.com/beehive-lab/TornadoVM/files/4825831/testcase_git.zip)

**Comment by jjfumero** [comment]
I suggest you create a repository with this use case, so it is easy to comment on changes. 

I did minor modifications, you don't need all data stream-in, since you only modify a two of the arrays:


```java
TaskSchedule ts = new TaskSchedule("Inner-Loop-Internal-UB-BCM17")
	.streamIn(outputDP, v_k) 
	.task("taskBCM17-"+v_k[0],TestCase::parallelInternalBCM17,inputDP,outputDP,weights,profits,v_cap,v_k) 
	.streamOut(outputDP); 
	
	for (int k = count; k < size; k++) {
		// Update k value of array
		v_k[0] = k;
			
		// Sequential execution
		parallelInternalBCM17(seqInputDP,seqOutputDP,weights,profits,v_cap,v_k);
		
		// Execution of taskschedule		
		ts.execute();
		
		// Update input arrays
		//inputDP = outputDP.clone();    // < don't clone
		//seqInputDP = seqOutputDP.clone();    // < don't clone
		
		// Print debug information
		System.out.println("Iteration/item " +k+". TaskSchedule output: " +outputDP[cap]+"; Sequential (correct) output: " +seqOutputDP[cap]);
```

I am getting this output

```java
Iteration/item 116. TaskSchedule output: 6.0; Sequential (correct) output: 6.0
task info: Inner-Loop-Internal-UB-BCM17.taskBCM17-0
	platform          : NVIDIA CUDA
	device            : GeForce GTX 1050 CL_DEVICE_TYPE_GPU (available)
	dims              : 0
	global work offset: [0]
	global work size  : [1]
	local  work size  : null
Iteration/item 117. TaskSchedule output: 4.0; Sequential (correct) output: 4.0
task info: Inner-Loop-Internal-UB-BCM17.taskBCM17-0
	platform          : NVIDIA CUDA
	device            : GeForce GTX 1050 CL_DEVICE_TYPE_GPU (available)
	dims              : 0
	global work offset: [0]
	global work size  : [1]
	local  work size  : null
Iteration/item 118. TaskSchedule output: 1.0; Sequential (correct) output: 1.0
task info: Inner-Loop-Internal-UB-BCM17.taskBCM17-0
	platform          : NVIDIA CUDA
	device            : GeForce GTX 1050 CL_DEVICE_TYPE_GPU (available)
	dims              : 0
	global work offset: [0]
	global work size  : [1]
	local  work size  : null
Iteration/item 119. TaskSchedule output: 2.0; Sequential (correct) output: 2.0
```

It seems to be correct. 


As I already commented, the kernel is being compiled and executed with OpenCL. However, it is a sequential kernel due to a bug in TornadoVM. We will work on that. As a work-around, can you change the loop-bound of the parallel loop in this manner?:


```java
for (@Parallel int c = 0; c <= input.length; c++){ 
   ... 
}
```

The problem is the constant as a loop-bound. Note also that if the value is to low, TornadoVM unrolls the whole loop.

**Comment by vtschouten97** [comment]
The output you show is incorrect in the context of having the desired value as output. But taking your adjustments into account and modifying the code further, I think it works now! Cloning of the array was likely to be the culprit, but it is necessary to have the outputDP function as an input in the following iteration, so I've played around a bit. I've added the final version of the TestCase project, in case anyone is itnterested. Now I only need to make sure it works in the context of my branch-and-bound algorithm.

The debug flag now shows the following:
task info: Inner-Loop-Internal-UB-BCM17-odd.taskBCM17
	platform          : NVIDIA CUDA
	device            : GeForce RTX 2060 CL_DEVICE_TYPE_GPU (available)
	dims              : 1
	global work offset: [0]
	global work size  : [1001]
	local  work size  : [1001, 1, 1]
Iteration/item 499. TaskSchedule output: 299.0; Sequential (correct) output: 299.0
task info: Inner-Loop-Internal-UB-BCM17-even.taskBCM17
	platform          : NVIDIA CUDA
	device            : GeForce RTX 2060 CL_DEVICE_TYPE_GPU (available)
	dims              : 1
	global work offset: [0]
	global work size  : [1001]
	local  work size  : [1001, 1, 1]
Iteration/item 500. TaskSchedule output: 299.0; Sequential (correct) output: 299.0
Final output: 299.0

[TestCase_final.zip](https://github.com/beehive-lab/TornadoVM/files/4830606/TestCase_final.zip)

**Comment by vtschouten97** [comment]
I haven't thought about posting here since I managed to get everything working properly, but I would like to thank you for the support you all have given me. In case you're interested in a branch-and-bound algorithm implemented with TornadoVM, I've added my thesis to this message. Cheers!

[Bachelor_Thesis_449249_FINAL.pdf](https://github.com/beehive-lab/TornadoVM/files/5163697/Bachelor_Thesis_449249_FINAL.pdf)
